### ISSUE TITLE
BigQuery: make jobs awaitable

### ISSUE THREAD

**Issue by dkapitan** [googleapis/google-cloud-python#15742]
I know BigQuery jobs are asynchronous by default. However, I am struggling to make my datapipeline async end-to-end.

Looking at [this JS example](https://medium.com/skyshidigital/what-bigquery-is-and-isnt-d092fa173d53), I thought it would be the most Pythonic to make a BigQuery job awaitable. However, I can't get that to work in Python i.e. errors when `await client.query(query)`. Looking at the source code, I don't see which method [returns an awaitable object](https://www.python.org/dev/peps/pep-0492/).

 I have little experience in writing async Python code and found [this example that wraps jobs in a `async def coroutine`](https://stackoverflow.com/questions/53065340/python-how-to-big-query-asynchronous-tasks).
```
class BQApi(object):                                                                                                 
    def __init__(self):                                                                                              
        self.api = bigquery.Client.from_service_account_json(BQ_CONFIG["credentials"])                               

    async def exec_query(self, query, **kwargs) -> bigquery.table.RowIterator:                                       
        job = self.api.query(query, **kwargs)                                                                        
        task = asyncio.create_task(self.coroutine_job(job))                                                          
        return await task                                                                                            

    @staticmethod                                                                                                    
    async def coroutine_job(job):                                                                                    
        return job.result()   
```

The [`google.api_core.operation.Operation`](https://googleapis.github.io/google-cloud-python/latest/core/operation.html) shows how to use `add_done_callback` to asynchronously wait for long-running operations. I have tried that, but the following yields `AttributeError: 'QueryJob' object has no attribute '_condition'` :

```
from concurrent.futures import ThreadPoolExecutor, as_completed
query1 = 'SELECT 1'
query2 = 'SELECT 2'

def my_callback(future):
    result = future.result()

operations = [bq.query(query1), bq.query(query2)]
[operation.add_done_callback(my_callback) for operation in operations]
results2 = []
for future in as_completed(operations):
  results2.append(list(future.result()))
```

Given that jobs are already asynchronous, would it make sense to add a method that returns an awaitable?

Or am I missing something and is there an Pythonic way to use the BigQuery client with the async/await pattern?

**Comment by RomeoAva** [comment]
Bump on this again :(

**Comment by chamini2** [comment]
WE NEED THIS

**Comment by niulai6** [comment]
WE NEED THIS!!!

**Comment by BnnaFish** [comment]
WE NEED THIS!!!

even adk-python is async-first

**Comment by bhuesemann** [comment]
This is 6 years old now. It will not come. There is no systematic support for async, and it seems to have no priority.

@kiraksi Have you heard anything about future plans for this?

**Comment by tswast** [comment]
There has been work to integrate BigQuery V2 APIs into https://github.com/googleapis/googleapis/tree/master/google/cloud/bigquery/v2

Theoretically, this could unlock the use of https://github.com/googleapis/gapic-generator-python which supports creating async clients.

CC @shollyman and @chalmerlowe who have been pushing forward on many of these efforts. IMO even the low-level generated clients could be useful for such use cases.

**Comment by bhuesemann** [comment]
> There has been work to integrate BigQuery V2 APIs into https://github.com/googleapis/googleapis/tree/master/google/cloud/bigquery/v2
> 
> Theoretically, this could unlock the use of https://github.com/googleapis/gapic-generator-python which supports creating async clients.
> 
> CC [@shollyman](https://github.com/shollyman) and [@chalmerlowe](https://github.com/chalmerlowe) who have been pushing forward on many of these efforts. IMO even the low-level generated clients could be useful for such use cases.

@tswast Thanks for giving some feedback and reigniting hope that this topic might be resolved in the future.

**Comment by bhuesemann** [comment]
> > There has been work to integrate BigQuery V2 APIs into https://github.com/googleapis/googleapis/tree/master/google/cloud/bigquery/v2
> > Theoretically, this could unlock the use of https://github.com/googleapis/gapic-generator-python which supports creating async clients.
> > CC [@shollyman](https://github.com/shollyman) and [@chalmerlowe](https://github.com/chalmerlowe) who have been pushing forward on many of these efforts. IMO even the low-level generated clients could be useful for such use cases.
> 
> [@tswast](https://github.com/tswast) Thanks for giving some feedback and reigniting hope that this topic might be resolved in the future.

Another month without transparency, no serious expectation management, no roadmap. Reignation failed ... putting this to sleep for another some years... Right? @tswast @shollyman @chalmerlowe

**Comment by tswast** [comment]
@bhuesemann I definitely understand the frustration. Async Python is not a "new" feature anymore. Thanks for the reminder. I'll see what I can do to push this forward.

**Comment by bhuesemann** [comment]
@tswast Thanks for your reply. I appreciate your efforts to move this foreward. But to be honest: I've lost hope that this will get priority (if not as part of something else). For me async processing is such a basic, essential thing especially for these naturally IO bound tasks. It keeps complexity low and provides easy scalability with minimal overhead. It should be imperative for such kind/type of APIs and the natural design choice. Thanks Tim, I know it can be sometimes hard to convice people to invest in such things. Please try what you can do as you said.

**Comment by Tunneller** [comment]
I’m going the 3.14t route.

 

I have Cloud Run running free threaded and the browser ends up triggering ~3 parallel (nonoverlapping) queries to BQ, but for the download I get iterative PyArrow pages and I send each page off to a thread. So this is more-or-less async. Maybe better.

 

Regards,

**Comment by shelbylsmith** [comment]
Pretty please address this issue. The difference async makes in performance/usability is massive.

**Comment by Tunneller** [comment]
Is this still pertinent now that Python has released the GIL ? I'm running 3.14t noGIL on Cloud Run with individual threads downloading from BQ and everything is in parallel. No async within the thread, but hey, I have 8 threads. And yes, much faster.

**Comment by shelbylsmith** [comment]
@Tunneller Yes it is pertinent or they would simply remove asyncio from python after the GIL changes. The GIL wasn't the bottleneck because the operations are mostly IO bound; nearly all IO operations actually release the GIL while doing syscalls. 8 threads means 8 concurrent requests maximum. That's pretty bad performance if they're doing a lot of waiting. Suppose you needed to run 8 queries in parallel to service a request, your container instance can only handle one request a time, which might be compounded if you need to thread things other than bigquery queries. Combine this with the fact that you don't have to fumble with threads or be concerned about thread-safety.

**Comment by Tunneller** [comment]
Ah yes, I had missed that it is not uncommon for asyncio users to launch hundreds of calls. You make a good point that threads are a finite and relatively expensive resource, and if you're using them for I/O waiting you're squandering them, even if they are (now) parallel.

**Comment by bhuesemann** [comment]
Please note that async, multi-processing and multi-threading are different concepts and relate to different use-cases for parallel processing. This use-case (db client query/processing) is inherently an IO-bound task (e.g. network IO-waits) and thus the typical use-case for async. Using a different concept for this is simply wrong and introduces unnecessary overhead and complexity (like handling concurrency).

**Comment by Tunneller** [comment]
BigQuery splits large downloads into multiple parallel streams via the Storage Read API, where each stream is a potentially blocking/synchronous operation, so threading is not inappropriate — we're not waiting on a single IO response but managing multiple concurrent blocking downloads, which is exactly the use-case for a thread pool. But I did not mean to imply no role for async.

Changing subject, slightly, I saw this: gcloud-aio-bigquery.  Could that offer faster downloads?

**Comment by bhuesemann** [comment]
Take a look here for an explanation: [https://stackoverflow.com/a/79619730](https://stackoverflow.com/a/79619730). Having true async support (also async iterators/generators) we could process multiple queries and their results in parallel on a single thread on the client using coroutines. gcloud-aio-bigquery seems to be a good async alternative to execute and process queries. Using the Google Storage API you have already async support on the client side (but a different use case).

**Comment by Tunneller** [comment]
Sure, but this really needs to be a request to BQ guys. I am but a humble user.

**Comment by chalmerlowe** [comment]
This issue was moved from the following repo as part of the monorepo migration: googleapis/python-bigquery
